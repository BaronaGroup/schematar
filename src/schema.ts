import ObjectId from './object-id'
import Complex from './complex'

export type DefaultContext = 'mongoose' | 'typescript' | 'jsonschema'

export interface Schema {
    fields: SchemaFields
}

type SimpleType = Symbol | Complex | typeof String | typeof Number | typeof Boolean | typeof Date | typeof Object
type IndexOptions = false | true | 'unique' | 'unique-sparse'

export type PlainType = SimpleType | [FieldInfo | SimpleType]

export type Field = FieldInfo | PlainType

export interface FieldInfo {
    type: PlainType
    index?: IndexOptions
    format?: any
    optional?: boolean
    presentIn?: Array<string>
    optionalIn?: Array<string>
    enum?: string[]
    mongooseDefault?: any
    mongooseRef?: string
}

export interface SchemaFields {
    [name: string]: Field
}
