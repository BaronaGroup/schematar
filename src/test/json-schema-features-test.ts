import Complex from '../complex'
import { Schema } from '../schema'
import { testJSONSchema } from './test-utils'

describe('json-schema-features', function () {
  const simpleSchema: Schema = {
    fields: {
      field: String,
      field2: Number,
    },
  }
  it(
    'allowNull',
    testJSONSchema({ fields: { field: { type: String, allowNull: true, optional: true } } }, (schema) => {
      expect(schema).toMatchSnapshot()
    })
  )

  it(
    'schemaId',
    testJSONSchema(
      { fields: { field: { type: String, allowNull: true, optional: true } } },
      (schema) => {
        expect(schema).toMatchSnapshot()
      },
      { schemaId: 'myschema' }
    )
  )

  it(
    'custom fields',
    testJSONSchema({ fields: { field: { type: String, jsonSchema: { year: 2020 } } } }, (schema) => {
      expect(schema).toMatchSnapshot()
    })
  )

  describe('makeEverythingOptional', function () {
    it(
      'shallow',
      testJSONSchema(
        simpleSchema,
        (schema) => {
          expect(schema).toMatchSnapshot()
        },
        { makeEverythingOptional: true }
      )
    )

    it(
      'deep',
      testJSONSchema(
        {
          fields: {
            topLevel: String,
            deep1: [
              new Complex({
                key: Number,
              }),
            ],
            deep2: new Complex({
              key2: Number,
              deep2: new Complex({
                key3: Number,
              }),
            }),
          },
        },
        (schema) => {
          expect(schema).toMatchSnapshot()
        },
        { makeEverythingOptional: true }
      )
    )
  })

  describe('allowAdditionalFields', function () {
    it(
      'shallow',
      testJSONSchema(
        simpleSchema,
        (schema) => {
          expect(schema).toMatchSnapshot()
        },
        { allowAdditionalFields: true }
      )
    )

    // Note: is not deep
    it(
      'deep',
      testJSONSchema(
        {
          fields: {
            topLevel: String,
            deep1: [
              new Complex({
                key: Number,
              }),
            ],
            deep2: new Complex({
              key2: Number,
              deep2: new Complex({
                key3: Number,
              }),
            }),
          },
        },
        (schema) => {
          expect(schema).toMatchSnapshot()
        },
        { allowAdditionalFields: true }
      )
    )
  })
})
