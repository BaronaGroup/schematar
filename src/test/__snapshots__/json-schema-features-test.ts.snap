// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`json-schema-features allowAdditionalFields deep 1`] = `
"{
  \\"type\\": \\"object\\",
  \\"$schema\\": \\"http://json-schema.org/draft-07/schema#\\",
  \\"properties\\": {
    \\"topLevel\\": {
      \\"type\\": \\"string\\"
    },
    \\"deep1\\": {
      \\"type\\": \\"array\\",
      \\"items\\": {
        \\"type\\": \\"object\\",
        \\"required\\": [
          \\"key\\"
        ],
        \\"properties\\": {
          \\"key\\": {
            \\"type\\": \\"number\\"
          }
        },
        \\"additionalProperties\\": false
      }
    },
    \\"deep2\\": {
      \\"type\\": \\"object\\",
      \\"required\\": [
        \\"key2\\",
        \\"deep2\\"
      ],
      \\"properties\\": {
        \\"key2\\": {
          \\"type\\": \\"number\\"
        },
        \\"deep2\\": {
          \\"type\\": \\"object\\",
          \\"required\\": [
            \\"key3\\"
          ],
          \\"properties\\": {
            \\"key3\\": {
              \\"type\\": \\"number\\"
            }
          },
          \\"additionalProperties\\": false
        }
      },
      \\"additionalProperties\\": false
    }
  },
  \\"required\\": [
    \\"topLevel\\",
    \\"deep1\\",
    \\"deep2\\"
  ],
  \\"additionalProperties\\": true
}"
`;

exports[`json-schema-features allowAdditionalFields shallow 1`] = `
"{
  \\"type\\": \\"object\\",
  \\"$schema\\": \\"http://json-schema.org/draft-07/schema#\\",
  \\"properties\\": {
    \\"field\\": {
      \\"type\\": \\"string\\"
    },
    \\"field2\\": {
      \\"type\\": \\"number\\"
    }
  },
  \\"required\\": [
    \\"field\\",
    \\"field2\\"
  ],
  \\"additionalProperties\\": true
}"
`;

exports[`json-schema-features allowNull 1`] = `
"{
  \\"type\\": \\"object\\",
  \\"$schema\\": \\"http://json-schema.org/draft-07/schema#\\",
  \\"properties\\": {
    \\"field\\": {
      \\"type\\": [
        \\"string\\",
        \\"null\\"
      ]
    }
  },
  \\"required\\": []
}"
`;

exports[`json-schema-features custom fields 1`] = `
"{
  \\"type\\": \\"object\\",
  \\"$schema\\": \\"http://json-schema.org/draft-07/schema#\\",
  \\"properties\\": {
    \\"field\\": {
      \\"type\\": \\"string\\",
      \\"year\\": 2020
    }
  },
  \\"required\\": [
    \\"field\\"
  ]
}"
`;

exports[`json-schema-features makeEverythingOptional deep 1`] = `
"{
  \\"type\\": \\"object\\",
  \\"$schema\\": \\"http://json-schema.org/draft-07/schema#\\",
  \\"properties\\": {
    \\"topLevel\\": {
      \\"type\\": \\"string\\"
    },
    \\"deep1\\": {
      \\"type\\": \\"array\\",
      \\"items\\": {
        \\"type\\": \\"object\\",
        \\"required\\": [],
        \\"properties\\": {
          \\"key\\": {
            \\"type\\": \\"number\\"
          }
        },
        \\"additionalProperties\\": false
      }
    },
    \\"deep2\\": {
      \\"type\\": \\"object\\",
      \\"required\\": [],
      \\"properties\\": {
        \\"key2\\": {
          \\"type\\": \\"number\\"
        },
        \\"deep2\\": {
          \\"type\\": \\"object\\",
          \\"required\\": [],
          \\"properties\\": {
            \\"key3\\": {
              \\"type\\": \\"number\\"
            }
          },
          \\"additionalProperties\\": false
        }
      },
      \\"additionalProperties\\": false
    }
  },
  \\"required\\": []
}"
`;

exports[`json-schema-features makeEverythingOptional shallow 1`] = `
"{
  \\"type\\": \\"object\\",
  \\"$schema\\": \\"http://json-schema.org/draft-07/schema#\\",
  \\"properties\\": {
    \\"field\\": {
      \\"type\\": \\"string\\"
    },
    \\"field2\\": {
      \\"type\\": \\"number\\"
    }
  },
  \\"required\\": []
}"
`;

exports[`json-schema-features schemaId 1`] = `
"{
  \\"$id\\": \\"myschema\\",
  \\"type\\": \\"object\\",
  \\"$schema\\": \\"http://json-schema.org/draft-07/schema#\\",
  \\"properties\\": {
    \\"field\\": {
      \\"type\\": [
        \\"string\\",
        \\"null\\"
      ]
    }
  },
  \\"required\\": []
}"
`;
